--- orig/include/sepol/mls.h
+++ mod/include/sepol/mls.h
@@ -15,19 +15,23 @@
 
 #ifdef CONFIG_SECURITY_SELINUX_MLS
 
-void mls_compute_av(context_struct_t * scontext,
+void mls_compute_av(policydb_t *policydb,
+		    context_struct_t * scontext,
 		    context_struct_t * tcontext,
 		    class_datum_t * tclass,
 		    access_vector_t * allowed);
 
-int mls_compute_context_len(context_struct_t * context);
+int mls_compute_context_len(policydb_t *policydb,
+			    context_struct_t * context);
 
-int mls_sid_to_context(context_struct_t * context,
+int mls_sid_to_context(policydb_t *policydb,
+		       context_struct_t * context,
 		       char **scontext);
 
 int mls_context_isvalid(policydb_t *p, context_struct_t * c);
 
-int mls_context_to_sid(char oldc,
+int mls_context_to_sid(policydb_t *policydb,
+		       char oldc,
 	               char **scontext,
 		       context_struct_t * context);
 
@@ -35,7 +39,8 @@
 			policydb_t * newp,
 			context_struct_t * context);
 
-int mls_compute_sid(context_struct_t *scontext,
+int mls_compute_sid(policydb_t *policydb,
+		    context_struct_t *scontext,
 		    context_struct_t *tcontext,
 		    security_class_t tclass,
 		    uint32_t specified,
@@ -45,8 +50,8 @@
 int cat_index(hashtab_key_t key, hashtab_datum_t datum, void *datap);
 int sens_destroy(hashtab_key_t key, hashtab_datum_t datum, void *p);
 int cat_destroy(hashtab_key_t key, hashtab_datum_t datum, void *p);
-int sens_read(policydb_t * p, hashtab_t h, void * fp);
-int cat_read(policydb_t * p, hashtab_t h, void * fp);
+int sens_read(policydb_t * p, hashtab_t h, struct policy_file *fp);
+int cat_read(policydb_t * p, hashtab_t h, struct policy_file *fp);
 
 #define mls_for_user_ranges(user, usercon) { \
 mls_range_list_t *__ranges; \
@@ -75,13 +80,13 @@
 
 #else
 
-#define	mls_compute_av(scontext, tcontext, tclass_datum, allowed)
-#define mls_compute_context_len(context) 0
-#define	mls_sid_to_context(context, scontextpp)
+#define	mls_compute_av(policydb, scontext, tcontext, tclass_datum, allowed)
+#define mls_compute_context_len(policydb, context) 0
+#define	mls_sid_to_context(policydb, context, scontextpp)
 #define mls_context_isvalid(p, c) 1
-#define	mls_context_to_sid(oldc, context_str, context) 0
+#define	mls_context_to_sid(policydb, oldc, context_str, context) 0
 #define mls_convert_context(oldp, newp, c) 0
-#define mls_compute_sid(scontext, tcontext, tclass, specified, newcontextp) 0
+#define mls_compute_sid(policydb, scontext, tcontext, tclass, specified, newcontextp) 0
 #define mls_for_user_ranges(user, usercon) 
 #define mls_end_user_ranges
 #define mls_symtab_names
